arch/riscv/core/thread.c

```diff
@@ -201,6 +202,15 @@ int arch_float_enable(struct k_thread *thread, unsigned int options)
 }
 #endif /* CONFIG_FPU && CONFIG_FPU_SHARING */
 
+#ifdef CONFIG_THREAD_LOCAL_STORAGE
+void arch_switch_to_main_thread(struct k_thread *main_thread, char *stack_ptr,
+				k_thread_entry_t _main)
+{
+	__asm__("mv tp, %0" : : "r" (main_thread->tls));
+	z_swap_unlocked();
```

Comment:
Non-main threads go through a regular context switch when they launch, and that code sets the tp register. The main thread doesn't go through that path. I don't think there's a common code path where this would naturally get set in both cases.

Reply:
But z_swap() is the regular context switch, or the front end API to it (some architectures implement it directly, others use an arch_switch() at a lower level).  And it's responsible for setting that register, like all registers, from wherever it's stored in the context.  Seems like we should just put this value there somewhere along the initialization path.

Reply:
Turns out the tracing code was getting invoked before the registers were set up and using k_current_get. And, of course, this happened before z_thread_entry was invoked, so you'd also not have z_tls_current set. The fix is simple, use z_current_get in these two functions. I've pushed a new patch set with this change.